resolve.c

```diff
@@ -62,68 +70,93 @@ char *resolveHostname(const char *addr)
 	return hostname;
 }
 
-// This routine is run *after* garbage cleaning (default interval is once per hour)
-// to account for possibly updated hostnames
-void reresolveHostnames(void)
+// Resolve client host names
+void resolveClients(bool onlynew)
 {
-	int clientID;
-	for(clientID = 0; clientID < counters.clients; clientID++)
+	int i;
+	for(i = 0; i < counters.clients; i++)
 	{
 		// Memory validation
-		validate_access("clients", clientID, true, __LINE__, __FUNCTION__, __FILE__);
-
-		// Process this client only if it has at least one active query in the log
-		if(clients[clientID].count < 1)
-			continue;
+		validate_access("clients", i, true, __LINE__, __FUNCTION__, __FILE__);
 
-		// Get client hostname
-		char *hostname = resolveHostname(clients[clientID].ip);
-		if(strlen(hostname) > 0)
+		// If onlynew flag is set, we will only resolve new clients
+		// If not, we will try to re-resolve all known clients
+		if(onlynew && clients[i].new)
 		{
-			// Delete possibly already existing hostname pointer before storing new data
-			if(clients[clientID].name != NULL)
-			{
-				free(clients[clientID].name);
-				clients[clientID].name = NULL;
-			}
-
-			// Store client hostname
-			clients[clientID].name = strdup(hostname);
-			clients[clientID].new = false;
+			char *hostname = resolveHostname(clients[i].ip);
+
+			enable_thread_lock();
+
+			if(clients[i].name != NULL)
+				free(clients[i].name);
+
+			clients[i].name = hostname;
+			clients[i].new = false;
+
+			disable_thread_lock();
 		}
-		free(hostname);
 	}
 }
 
-// This routine is run *before* saving to the database (default interval is once per minute)
-// to account for new clients (and forward destinations)
-void resolveNewClients(void)
+// Resolve upstream destination host names
+void resolveForwardDestinations(bool onlynew)
 {
 	int i;
-	for(i = 0; i < counters.clients; i++)
+	for(i = 0; i < counters.forwarded; i++)
 	{
 		// Memory validation
-		validate_access("clients", i, true, __LINE__, __FUNCTION__, __FILE__);
+		validate_access("forwarded", i, true, __LINE__, __FUNCTION__, __FILE__);
 
-		// Only try to resolve new clients
-		// Note that it can happen that we are not able to find hostnames but we don't
-		// want to try to resolve them every minute in this case.
-		if(clients[i].new)
+		// If onlynew flag is set, we will only resolve new upstream destinations
+		// If not, we will try to re-resolve all known upstream destinations
+		if(onlynew && forwarded[i].new)
 		{
-			clients[i].name = resolveHostname(clients[i].ip);
-			clients[i].new = false;
+			char *hostname = resolveHostname(forwarded[i].ip);
+
+			enable_thread_lock();
+
+			if(forwarded[i].name != NULL)
+				free(forwarded[i].name);
+
+			forwarded[i].name = hostname;
+			forwarded[i].new = false;
+
+			disable_thread_lock();
 		}
 	}
-	for(i = 0; i < counters.forwarded; i++)
+}
+
+pthread_t resolverthread;
```

Comment:
Can this be moved to where the other threads are defined (main)?

Reply:
They are all defined in `dnsmasq_interface.c` (close to where the threads are launched). I'll move the definition there.

Reply:
Ah ya, sorry, thought that was main for a second (we used to launch the threads from there).